File: tbco/ws-api-client.py
import asyncio
import requests
import websockets
from datetime import datetime

ENDPOINT = "wss://www.theblock.co/ws"
# ENDPOINT = "ws://localhost:4000/ws"

LAST_PING_TIMEOUT = 60000


async def read_api_key():
    try:
        with open("./ws-api-key.txt", "r") as file:
            key = file.read().strip()
            if not key:
                raise Exception("Empty key file")
            return key
    except Exception:
        raise Exception("API key file not found or empty")


def current_time():
    return datetime.now().strftime("%Y-%m-%d %H:%M:%S")


async def connect():
    last_ping = datetime.now().timestamp() * 1000

    key = await read_api_key()

    async with websockets.connect(ENDPOINT, extra_headers={"x-auth-token": key}) as ws:
        print("Connected to", ENDPOINT)
        while True:
            try:
                message = await ws.recv()
                print(f"[{current_time()}] PAYLOAD:", message)

            except websockets.exceptions.ConnectionClosed as e:
                print("ERROR:", e)
                await asyncio.sleep(10)
                await connect()


asyncio.run(connect())

File: tbco/ws-api-client.js
const WebSocket = require("ws");
const fs = require("fs");
const { DateTime } = require("luxon");

const ENDPOINT = "wss://www.theblock.co/ws";
// const ENDPOINT = 'ws://localhost:4000/ws';

const LAST_PING_TIMEOUT = 60000;
const PING_INTERVAL = 25000; // 25 seconds

async function readApiKey() {
  try {
    const key = fs.readFileSync("./ws-api-key.txt", "utf8").trim();
    if (!key) {
      throw new Error("Empty key file");
    }
    return key;
  } catch (error) {
    throw new Error("API key file not found or empty");
  }
}

function currentTime() {
  return DateTime.now().toFormat("yyyy-MM-dd HH:mm:ss");
}

async function connect() {
  let lastPing = DateTime.local().toMillis();

  const key = await readApiKey();

  const ws = new WebSocket(ENDPOINT, {
    headers: {
      "x-auth-token": key,
    },
  });

  ws.on("open", () => {
    console.log("Connected to", ENDPOINT);
    // Start the ping interval
    setInterval(() => {
      if (DateTime.local().toMillis() - lastPing > LAST_PING_TIMEOUT) {
        console.log("Ping timeout, reconnecting...");
        ws.terminate(); // Close the connection
        connect(); // Reconnect
      } else {
        ws.ping(); // Send ping
      }
    }, PING_INTERVAL);
  });

  ws.on("message", (message) => {
    console.log(`[${currentTime()}] PAYLOAD:`, message.toString());
  });

  ws.on("close", function close() {
    console.log("Connection closed");
    setTimeout(connect, 10000);
  });

  ws.on("error", function error(err) {
    console.error("ERROR:", err.message);
    setTimeout(connect, 10000);
  });

  ws.on("pong", function pong() {
    lastPing = DateTime.local().toMillis();
    console.log(`[${currentTime()}] pong`);
  });
}

connect();
